gpu-simulator/trace-parser/trace_parser.cc

```diff
@@ -289,67 +290,12 @@ kernel_trace_t *trace_parser::parse_kernel_info(
   kernel_info->enable_lineinfo = 0;  // default disabled
 
   std::string read_trace_cmd;
-  int _l = kerneltraces_filepath.length();
-  if (_l > 3 && kerneltraces_filepath.substr(_l - 3, 3) == ".xz") {
-    // this is xz-compressed trace
-    read_trace_cmd = "xz -dc " + kerneltraces_filepath;
-  } else if (_l > 7 && kerneltraces_filepath.substr(_l - 7, 7) == ".traceg") {
-    // this is plain text trace
-    read_trace_cmd = "cat " + kerneltraces_filepath;
-  } else {
-    std::cerr << "Can't read trace. Only .xz and plain text are supported: "
-              << kerneltraces_filepath << "\n";
-    exit(1);
-  }
-
-  // Create an interprocess channel, and fork out a data source process. The
-  // data source process reads trace from disk, write to the channel, and the
-  // simulator process read from the channel.
-  int *pipefd = kernel_info->pipefd;
-  if (pipe(pipefd) != 0) {
-    std::cerr << "Failed to create interprocess channel\n";
-    perror("pipe");
-    exit(1);
-  }
-
-  pid_t pid = fork();
-  if (pid == 0) {
-    // The child process is the data source. Redirect its
-    // stdout to the write end of the pipe.
-    close(pipefd[0]);
-    dup2(pipefd[1], STDOUT_FILENO);
-
-    // When using GDB, sending Ctrl+C to the simulator will send a SIGINT signal
-    // to the child process as well, subsequently causing it to terminate. To
-    // avoid this, we let the child process ignore (SIG_IGN) the SIGINT signal.
-    // Reference:
-    // https://stackoverflow.com/questions/38404925/gdb-interrupt-running-process-without-killing-child-processes
-    signal(SIGINT, SIG_IGN);
-
-    execle("/bin/sh", "sh", "-c", read_trace_cmd.c_str(), NULL, environ);
-    perror("execle");  // the child process shouldn't reach here if all is well.
-    exit(1);
-  } else {
-    // parent (simulator)
-    close(pipefd[1]);
-    dup2(pipefd[0], STDIN_FILENO);
-  }
-
-  // Parent continues from here.
-  kernel_info->ifs = &std::cin;
-  std::istream *ifs = kernel_info->ifs;
+  kernel_info->pipeReader.OpenFile(kerneltraces_filepath);
```

Comment:
Please see my comment for `class PipeReader` - I think it's better to merge the code of `OpenFile` into the ctor of `kernel_trace_t`.